File: 622-design-circular-queue/622-design-circular-queue.js
/**
 * @param {number} k
 */
let q;
let f=-1,r=-1;
let size=0,arrsize=0;
var MyCircularQueue = function(k) {
    q = new Array(k);
    arrsize = k;
    f=-1;r=-1;
    size=0;
};

/** 
 * @param {number} value
 * @return {boolean}
 */
MyCircularQueue.prototype.enQueue = function(value) {
    if(MyCircularQueue.prototype.isFull()) return false;
    
    if(MyCircularQueue.prototype.isEmpty()){
        f=0;r=0;
        q[r] = value;
        size++;
        return true;
    }
    
    r++;
    r%=arrsize;
    size++;
    q[r] = value;
    return true;
};

/**
 * @return {boolean}
 */
MyCircularQueue.prototype.deQueue = function() {
    if(MyCircularQueue.prototype.isEmpty()) return false;
    f++;
    f%=arrsize;
    size--;
    return true;
};

/**
 * @return {number}
 */
MyCircularQueue.prototype.Front = function() {
    if(size == 0) return -1;
    return q[f];
};

/**
 * @return {number}
 */
MyCircularQueue.prototype.Rear = function() {
    if(size == 0) return -1;
    return q[r];  
};

/**
 * @return {boolean}
 */
MyCircularQueue.prototype.isEmpty = function() {
    return size == 0;
};

/**
 * @return {boolean}
 */
MyCircularQueue.prototype.isFull = function() {
    return size == arrsize;
};

/** 
 * Your MyCircularQueue object will be instantiated and called as such:
 * var obj = new MyCircularQueue(k)
 * var param_1 = obj.enQueue(value)
 * var param_2 = obj.deQueue()
 * var param_3 = obj.Front()
 * var param_4 = obj.Rear()
 * var param_5 = obj.isEmpty()
 * var param_6 = obj.isFull()
 */

File: 622-design-circular-queue/622-design-circular-queue.cpp
class MyCircularQueue {
public:
    vector<int> q;
    int f=-1,r=-1;
    int size=0,arrsize=0;
    MyCircularQueue(int k) {
        q.resize(k);
        arrsize = k;
    }
    
    bool enQueue(int value) {
        if(isFull()) return 0;
        if(isEmpty()){
            f=0;r=0;
            q[r] = value;
            size++;
            return 1;
        }
        size++;
        r++;
        r%=arrsize;
        q[r] = value;
        return 1;
    }
    
    bool deQueue() {
        if(isEmpty()) return 0;
        
        f++;f%=arrsize;
        size--;
        return 1;
    }
    
    int Front() {
        if(isEmpty()) return -1;
        return q[f];   
    }
    
    int Rear() {
        if(isEmpty()) return -1;
        return q[r];
    }
    
    bool isEmpty() {
        return size == 0;
    }
    
    bool isFull() {
        return size == arrsize;
    }
};

/**
 * Your MyCircularQueue object will be instantiated and called as such:
 * MyCircularQueue* obj = new MyCircularQueue(k);
 * bool param_1 = obj->enQueue(value);
 * bool param_2 = obj->deQueue();
 * int param_3 = obj->Front();
 * int param_4 = obj->Rear();
 * bool param_5 = obj->isEmpty();
 * bool param_6 = obj->isFull();
 */